azure-ipam/ipam_test.go

```diff
@@ -0,0 +1,303 @@
+	args    *cniSkel.CmdArgs
+	wantErr bool
+}
+
+func TestCmdAdd(t *testing.T) {
+	req := require.New(t)
+
+	happyNetConf := &cniTypes.NetConf{
+		CNIVersion: "0.1.0",
+		Name:       "happynetconf",
+	}
+
+	invalidVersionNetConf := &cniTypes.NetConf{
+		CNIVersion: "0",
+		Name:       "nothappynetconf",
+	}
+
+	happyNetConfByteArr, err := json.Marshal(happyNetConf)
+	if err != nil {
+		panic(err)
+	}
+	invalidVersionNetConfByteArr, err := json.Marshal(invalidVersionNetConf)
+	if err != nil {
+		panic(err)
+	}
+	invalidNetConf := []byte("invalidNetConf")
+
+	happyArgs := &cniSkel.CmdArgs{
+		ContainerID: "happyArgs",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        happyPodArgs,
+		StdinData:   happyNetConfByteArr,
+	}
+
+	failCreateCNSReqArgs := &cniSkel.CmdArgs{
+		ContainerID: "failCreateCNSReqArgs",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        nothappyPodArgs,
+		StdinData:   happyNetConfByteArr,
+	}
+
+	failRequestCNSArgs := &cniSkel.CmdArgs{
+		ContainerID: "failRequestCNSArgs",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        happyPodArgs,
+		StdinData:   happyNetConfByteArr,
+	}
+
+	failProcessCNSResp := &cniSkel.CmdArgs{
+		ContainerID: "failProcessCNSResp",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        happyPodArgs,
+		StdinData:   happyNetConfByteArr,
+	}
+
+	failParseNetConf := &cniSkel.CmdArgs{
+		ContainerID: "failParseNetConf",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        happyPodArgs,
+		StdinData:   invalidNetConf,
+	}
+
+	failGetVersionedResult := &cniSkel.CmdArgs{
+		ContainerID: "failGetVersionedResult",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        happyPodArgs,
+		StdinData:   invalidVersionNetConfByteArr,
+	}
+
+	tests := []scenario{
+		{
+			name:    "Happy CNI add",
+			args:    happyArgs,
+			wantErr: false,
+		},
+		{
+			name:    "Fail create CNS request during CmdAdd",
+			args:    failCreateCNSReqArgs,
+			wantErr: true,
+		},
+		{
+			name:    "Fail request CNS ipconfig during CmdAdd",
+			args:    failRequestCNSArgs,
+			wantErr: true,
+		},
+		{
+			name:    "Fail process CNS response during CmdAdd",
+			args:    failProcessCNSResp,
+			wantErr: true,
+		},
+		{
+			name:    "Fail parse netconf during CmdAdd",
+			args:    failParseNetConf,
+			wantErr: true,
+		},
+		{
+			name:    "Fail get versioned result during CmdAdd",
+			args:    failGetVersionedResult,
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			fmt.Printf("test : %v\n", tt.name)
+			mockCNSClient := &MockCNSClient{}
+			logger, err := zap.NewProduction()
+			defer logger.Sync() // nolint
+			if err != nil {
+				return
+			}
+			ipamPlugin, _ := NewPlugin(logger, mockCNSClient)
+			err = ipamPlugin.CmdAdd(tt.args)
+			fmt.Println()
+			if tt.wantErr {
+				req.Error(err)
+			} else {
+				req.NoError(err)
+			}
```

Comment:
could we also validate the output? That's probably the most important thing -- need to make sure the IPs that CNS returns make it to the upper plugin.

I'm not totally sure how we can capture the output, but we might have examples of how we tested the other plugins?

Reply:
Here's an example of how to capture and verify the output (from the CNI reference plugins): https://github.com/containernetworking/plugins/blob/9c59728d39b729558bb712c79e2c80acc0864191/pkg/testutils/cmd.go#L42-L47

I think you can probably use the testutils package directly (it's part of the same library we're already using to invoke the delegated plugin)

Really think we should test this, because if something gets output incorrectly then the plugin won't work.